File: server/ssrf.ts
import dns from 'node:dns/promises';
import net from 'node:net';

// SSRF guard for server-side fetches of client-supplied URLs (e.g. /api/fetch-content,
// which on the public read-only demo is reachable anonymously). Resolves the host and
// refuses to fetch anything that points at a private / loopback / link-local / metadata
// address, so a visitor can't make the server probe the internal network.
//
// Known limitation: fetch() re-resolves DNS independently, so a determined attacker could
// rebind between this check and the fetch. Acceptable for a low-value home demo; pin the
// resolved IP if this ever guards something sensitive.

function ipv4ToInt(ip: string): number {
  const p = ip.split('.').map(Number);
  return ((p[0] << 24) >>> 0) + (p[1] << 16) + (p[2] << 8) + p[3];
}

function isPrivateIPv4(ip: string): boolean {
  const n = ipv4ToInt(ip);
  const inRange = (base: string, bits: number) =>
    n >>> (32 - bits) === ipv4ToInt(base) >>> (32 - bits);
  return (
    inRange('0.0.0.0', 8) || // "this" network
    inRange('10.0.0.0', 8) || // private
    inRange('100.64.0.0', 10) || // CGNAT
    inRange('127.0.0.0', 8) || // loopback
    inRange('169.254.0.0', 16) || // link-local (incl. 169.254.169.254 metadata)
    inRange('172.16.0.0', 12) || // private
    inRange('192.168.0.0', 16) // private
  );
}

// Expand any valid IPv6 string to its 8 numeric groups (handles `::` compression and a
// trailing embedded IPv4, e.g. `::ffff:1.2.3.4`). Returns null for non-IPv6 input.
function ipv6Groups(ip: string): number[] | null {
  if (!net.isIPv6(ip)) return null;
  let str = ip.toLowerCase();
  const v4 = str.match(/(\d+\.\d+\.\d+\.\d+)$/);
  if (v4) {
    const p = v4[1].split('.').map(Number);
    str =
      str.slice(0, v4.index) +
      ((p[0] << 8) | p[1]).toString(16) +
      ':' +
      ((p[2] << 8) | p[3]).toString(16);
  }
  const [head, tail] = str.split('::');
  const headGroups = head ? head.split(':') : [];
  const tailGroups = tail === undefined ? null : tail ? tail.split(':') : [];
  const groups =
    tailGroups === null
      ? headGroups
      : [
          ...headGroups,
          ...Array(8 - headGroups.length - tailGroups.length).fill('0'),
          ...tailGroups,
        ];
  if (groups.length !== 8) return null;
  return groups.map((g) => parseInt(g || '0', 16));
}

// Return the dotted IPv4 of an IPv4-mapped IPv6 address (`::ffff:a.b.c.d`, in any
// notation the URL parser / DNS may hand us), else null. Catches the hex-compressed
// form (`::ffff:7f00:1`) that a plain dotted-decimal regex would miss.
function mappedIPv4(ip: string): string | null {
  const g = ipv6Groups(ip);
  if (!g) return null;
  const highBitsZero = g[0] === 0 && g[1] === 0 && g[2] === 0 && g[3] === 0 && g[4] === 0;
  if (!highBitsZero || g[5] !== 0xffff) return null;
  return `${g[6] >> 8}.${g[6] & 0xff}.${g[7] >> 8}.${g[7] & 0xff}`;
}

function isPrivateIP(ip: string): boolean {
  const mapped = mappedIPv4(ip); // IPv4-mapped IPv6
  if (mapped) return isPrivateIPv4(mapped);
  if (net.isIPv4(ip)) return isPrivateIPv4(ip);
  const lower = ip.toLowerCase();
  if (lower === '::1' || lower === '::') return true; // loopback / unspecified
  const first = parseInt(lower.split(':')[0] || '0', 16);
  if ((first & 0xfe00) === 0xfc00) return true; // fc00::/7 unique-local
  if ((first & 0xffc0) === 0xfe80) return true; // fe80::/10 link-local
  return false;
}

/** Throws if `raw` is not an http(s) URL or resolves to a non-public address. */
export async function assertSafeUrl(raw: string): Promise<void> {
  let u: URL;
  try {
    u = new URL(raw);
  } catch {
    throw new Error('Invalid URL');
  }
  if (u.protocol !== 'http:' && u.protocol !== 'https:') {
    throw new Error('Only http/https URLs are allowed');
  }
  const host = u.hostname.replace(/^\[|\]$/g, ''); // strip IPv6 brackets
  if (net.isIP(host)) {
    if (isPrivateIP(host)) throw new Error('Blocked address');
    return;
  }
  const records = await dns.lookup(host, { all: true });
  if (records.length === 0) throw new Error('Host did not resolve');
  for (const { address } of records) {
    if (isPrivateIP(address)) throw new Error('Blocked address');
  }
}

File: server-go/internal/ssrf/ssrf.go
// Package ssrf is the Go port of server/ssrf.ts: an SSRF guard for server-side
// fetches of client-supplied URLs (/api/fetch-content). It resolves the host and
// refuses any URL that points at a private / loopback / link-local / metadata
// address, so a visitor can't make the server probe the internal network.
//
// Known limitation (same as the Node version): the HTTP client re-resolves DNS
// independently, so a determined attacker could rebind between this check and the
// fetch. Acceptable for a low-value home demo; pin the resolved IP if this ever
// guards something sensitive.
package ssrf

import (
	"context"
	"errors"
	"net"
	"net/url"
)

// privateV4CIDRs mirrors isPrivateIPv4 in ssrf.ts exactly (a superset of Go's
// net.IP.IsPrivate: it also covers 0.0.0.0/8, 100.64.0.0/10 CGNAT, all of
// 127/8 loopback, and 169.254/16 link-local incl. the cloud metadata endpoint).
var privateV4CIDRs = mustCIDRs(
	"0.0.0.0/8",      // "this" network
	"10.0.0.0/8",     // private
	"100.64.0.0/10",  // CGNAT
	"127.0.0.0/8",    // loopback
	"169.254.0.0/16", // link-local (incl. 169.254.169.254 metadata)
	"172.16.0.0/12",  // private
	"192.168.0.0/16", // private
)

func mustCIDRs(cidrs ...string) []*net.IPNet {
	out := make([]*net.IPNet, 0, len(cidrs))
	for _, c := range cidrs {
		_, n, err := net.ParseCIDR(c)
		if err != nil {
			panic(err)
		}
		out = append(out, n)
	}
	return out
}

func isPrivateIPv4(ip net.IP) bool {
	for _, n := range privateV4CIDRs {
		if n.Contains(ip) {
			return true
		}
	}
	return false
}

// isPrivateIP classifies any parsed IP. Go's net.ParseIP normalizes an
// IPv4-mapped IPv6 address (::ffff:a.b.c.d, in dotted or hex-compressed form) so
// that To4() yields the embedded IPv4 — covering the mapped cases ssrf.ts handles
// via mappedIPv4. Pure IPv6 is classified against the same ranges Node checks.
func isPrivateIP(ip net.IP) bool {
	if v4 := ip.To4(); v4 != nil {
		return isPrivateIPv4(v4)
	}
	// Pure IPv6: ::1 loopback, :: unspecified, fc00::/7 unique-local, fe80::/10
	// link-local. Go's helpers cover exactly these, matching isPrivateIP in ssrf.ts.
	return ip.IsLoopback() || ip.IsUnspecified() || ip.IsPrivate() || ip.IsLinkLocalUnicast()
}

// resolveHost is the DNS hook (overridable in tests via SetResolver). It returns
// every address the host resolves to, like dns.lookup(host, { all: true }).
var resolveHost = func(ctx context.Context, host string) ([]net.IP, error) {
	addrs, err := net.DefaultResolver.LookupIPAddr(ctx, host)
	if err != nil {
		return nil, err
	}
	ips := make([]net.IP, len(addrs))
	for i, a := range addrs {
		ips[i] = a.IP
	}
	return ips, nil
}

// AssertSafeURL returns an error if raw is not an http(s) URL or resolves to a
// non-public address — the port of assertSafeUrl. Error messages match ssrf.ts so
// the /api/fetch-content 400 detail is identical.
func AssertSafeURL(ctx context.Context, raw string) error {
	u, err := url.Parse(raw)
	// new URL(raw) throws on a schemeless / malformed input → "Invalid URL".
	if err != nil || u.Scheme == "" {
		return errors.New("Invalid URL")
	}
	if u.Scheme != "http" && u.Scheme != "https" {
		return errors.New("Only http/https URLs are allowed")
	}
	host := u.Hostname() // strips IPv6 brackets
	if host == "" {
		return errors.New("Invalid URL")
	}
	// Literal IP: classify directly, no DNS (matching the net.isIP branch).
	if ip := net.ParseIP(host); ip != nil {
		if isPrivateIP(ip) {
			return errors.New("Blocked address")
		}
		return nil
	}
	addrs, err := resolveHost(ctx, host)
	if err != nil {
		return errors.New("Host did not resolve")
	}
	if len(addrs) == 0 {
		return errors.New("Host did not resolve")
	}
	for _, ip := range addrs {
		if isPrivateIP(ip) {
			return errors.New("Blocked address")
		}
	}
	return nil
}
